### DISCUSSION TITLE
How to Achieve Field by Field Comparison Behaviour Without Recursively Traversing the Whole Object Graph

### DISCUSSION THREAD

**Discussion by raphaelNguyen** [assertj/assertj#3352] [Q&A]
Hi all,

```
A
 |— B b
 |    |— String s
 |    |— C c
 |         |— String s
 |         |— Date d
 |         |— [...] // complex object graph
 |— String i
 |— String j
 |— [...] // complex object graph
```

I have a use case where I want to assert equal between 2 objects (e.g. `A`) according to a subset of the fields (e.g. `i` and `j`) on the objects. These objects are part of an object graph so there are a number of fields that I don't want to recurse into for this assertion (e.g. `b`).

Previously, I was able to get this behaviour via `.usingRecursiveComparison().comparingOnlyFieldsOfType(String.class, [...])` and then specifying a few field names to ignore. It never recurse into the object graph and only assert equality using the fields I'm concerned with. However, #3207 specified that this was not intended behaviour and fixed it. With the new behaviour, the assertion takes significantly longer as it has to traverse the very complex object graph.

I've tried tweaking various settings in the recursive comparison configuration but have yet to be able to limit the traversal. I've also found that I might be able to achieve my desired behaviour with `.isEqualToComparingFieldByField()` but this have been deprecated in favour of `.usingRecursiveComparison()`.

My only recourse is if I extract all the fields I'm concerned with from the object and assert equality with them. This does have the drawback of not automatically taking into account new fields if I add them in the future. Without any better alternatives, this might be what I have to go with.

Can you please suggest if there might be a way for me to achieve the desired equality assertion behaviour with the current API?

Thank you very much for your time.

**Comment by Mackzwellz** [comment]
Consider taking the opposite approach - of ignoring the fields you don't need to compare.
AssertJ has method that allows to specify list of fields to ignore, but it's not really extensible, in my opinion, and can lead to a lot of repetitive / unnecessary code.

I have similar requirements for my project, and this is what I went with:

1. [Interface that allows to specify fields to ignore](https://github.com/Mackzwellz/assertj-list-comparison-examples/blob/master/src/main/java/io/github/mackzwellz/assertj/custom/FieldComparisonExcludable.java) - implement it in objects that you compare.

2. [Custom Introspection Strategy for AssertJ](https://github.com/Mackzwellz/assertj-list-comparison-examples/blob/master/src/test/java/assertj/custom/CustomIgnoringIntrospectionStrategy.java#L50-L57) that removes specified fields from comparison for each object - add it to each assertion's configuration, or create a wrapper assert / wrapper class for assertion.

**Reply by raphaelNguyen** [reply]
Thank you for your recommendations.
